datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java

```diff
@@ -130,6 +143,43 @@
     DYNAMIC_METRICS_MANAGER.registerGauge(_className, AGGREGATE, NUM_TOPICS, aggNumTopics::get);
   }
 
+  /**
+   * Register some of the Kafka consumer metrics of interest. This cannot be done as part of the constructor
+   * as the Kafka consumer may be created at a later time than this object.
+   * @param consumer the Kafka consumer for which to register the metrics
+   * @param clientId the Kafka consumer's client.id
+   */
+  public void registerKafkaConsumerMetrics(Consumer<?, ?> consumer, String clientId) {
+    if (consumer == null || StringUtils.isBlank(clientId)) {
+      _errorLogger.warn("Cannot register the Kafka consumer metrics, either the consumer is null or the client.id is blank");
+      return;
+    }
+
+    Supplier<Double> watermarkSpanSupplier = () -> getConsumerOffsetWatermarkSpanMetric(consumer, clientId);
+    DYNAMIC_METRICS_MANAGER.registerGauge(_className, _key, CONSUMER_OFFSET_WATERMARK_SPAN, watermarkSpanSupplier);
+
+    Supplier<Double> offsetResetSupplier = () -> getConsumerLiclosestDataLossEstimationMetric(consumer, clientId);
+    DYNAMIC_METRICS_MANAGER.registerGauge(_className, _key, CONSUMER_LICLOSEST_DATA_LOSS_ESTIMATION, offsetResetSupplier);
+
+    _kafkaConsumerMetricsRegistered = true;
+  }
+
+  private double getConsumerOffsetWatermarkSpanMetric(Consumer<?, ?> consumer, String clientId) {
+    return getConsumerMetricValue(consumer, clientId, "consumer-offset-watermark-span");
+
+  }
+
+  private double getConsumerLiclosestDataLossEstimationMetric(Consumer<?, ?> consumer, String clientId) {
+    return getConsumerMetricValue(consumer, clientId, "consumer-liclosest-data-loss-estimation");
+  }
+
+  private double getConsumerMetricValue(Consumer<?, ?> consumer, String clientId, String metricName) {
+    Map<String, String> tags = new HashMap<>(2);
+    tags.put("client-id", clientId);
+    MetricName name = new MetricName(metricName, "lnkd", "", tags);
+    return Optional.ofNullable(consumer).map(c -> c.metrics().get(name)).map(Metric::value).orElse(0.0);
```

Comment:
Are you deliberately using the deprecated `Metric::value` instead of `Metric.metricValue()` + casting?

Reply:
As discussed offline, my usage of the deprecated `Metric::value` was indeed deliberate. For using `Metric.metricValue()` + casting, you need to know the metric type, and unfortunately if Kafka changes the metric type underneath us, it can cause us to not emit the metric and there is no easy way to know why. On the other hand, if the deprecated method is removed or if the behavior is buggy, we have a problem too. Decided to go with the `Metric.metricValue()` + casting